### ISSUE TITLE
A new CRAN submission to solve the reported 'noLD' issue.

### ISSUE THREAD

**Issue by hetong007** [dmlc/xgboost#2960]
This was notified by **Prof Brian Ripley** through email:

> This concerns packages
> 
> CBPS ExtDist GGIR RInSp RSQLite drtmle expss gamlss.inf lakemorpho lhs partitionComparison pre recipes tableone vpc xgboost xpose
> 
> which have 'noLD' 'Additional issues' reported on their CRAN results pages.  And to quote the manual
> 
> <quote>
> Only test the accuracy of results if you have done a formal error analysis. Things such as checking that probabilities numerically sum to one are silly: numerical tests should always have a tolerance. That the tests on your platform achieve a particular tolerance says little about other platforms. R is configured by default to make use of long doubles where available, but they may not be available or be too slow for routine use. Most R platforms use ‘ix86’ or ‘x86_64’ CPUs: these use extended precision registers on some but not all of their FPU instructions. Thus the achieved precision can depend on the compiler version and optimization flags—our experience is that 32-bit builds tend to be less precise than 64-bit ones. But not all platforms use those CPUs, and not all which use them configure them to allow the use of extended precision. In particular, ARM CPUs do not (currently) have extended precision nor long doubles, and long double was 64-bit on HP/PA Linux.
> 
> If you must try to establish a tolerance empirically, configure and build R with --disable-long-double and use appropriate compiler flags (such as -ffloat-store and -fexcess-precision=standard for gcc, depending on the CPU type) to mitigate the effects of extended-precision calculations.
> </quote>
> 
> Packages partitionComparison and vpc do test for sums without a tolerance ....
> 
> Please submit a corrected update before **Jan 24** to safely retain the package on CRAN.

**Proposal**

As we are currently having some discussions about a new CRAN version (e.g.  https://github.com/dmlc/xgboost/issues/2948), it may be a good chance to work on a new version on CRAN. The difficulty for it mainly comes from https://github.com/dmlc/xgboost/issues/2319.

@khotilov @thirdwing, I would like to have your input on the following options:

1. Only fix the 'noLD' issue for the CRAN version and not introducing any changes on github.
2. A walkaround: fix the 'noLD' issue and merge our recent changes in R code to the CRAN version. However this may result in a different version from github.
3. Fix the 'noLD' issue and try again to solve https://github.com/dmlc/xgboost/issues/2319 . The benefit is obvious: this will sync our github repo and CRAN again, but the cause of the crash is still unclear.

Inputs from anyone or on new options are also welcome!

**Comment by khotilov** [comment]
Hopefully, #2994 would allow the (3).

An easy fix for the failing test in the noLD case would be to run that test only when `capabilities('long.double') == TRUE`. While it might seem like cheating, this is not an essential test and it only serves the purpose of verifying the assumptions about handling the double precision on most common platforms, and while it would be interesting to understand what is happening in the noLD build, it is not an urgent matter from my perspective.

**Comment by hetong007** [comment]
https://github.com/dmlc/xgboost/pull/3061 fixed the 'noLD' issue, however the current github version failed to pass the [auto test](http://win-builder.r-project.org) because of one of the following NOTE:

```
* checking CRAN incoming feasibility ... NOTE
Maintainer: 'Tong He <hetong007@gmail.com>'

Possibly mis-spelled words in DESCRIPTION:
  Guestrin (11:52)

...

* checking for GNU extensions in Makefiles ... NOTE
GNU make is a SystemRequirements.
* checking for portable use of $(BLAS_LIBS) and $(LAPACK_LIBS) ... OK
* checking compiled code ... NOTE
File 'xgboost/libs/i386/xgboost.dll':
  Found '_ZSt4cout', possibly from 'std::cout' (C++)
    Object: './amalgamation/xgboost-all0.o'
File 'xgboost/libs/x64/xgboost.dll':
  Found '_ZSt4cout', possibly from 'std::cout' (C++)
    Object: './amalgamation/xgboost-all0.o'

Compiled code should not call entry points which might terminate R nor
write to stdout/stderr instead of to the console, nor the system RNG.

See 'Writing portable packages' in the 'Writing R Extensions' manual.
```

The current package on CRAN is based on `0.6.4` with a quick fix of the 'noLD' issue, which corresponds to the first option. To submit a `0.7.0` version to CRAN as requested in https://github.com/dmlc/xgboost/issues/2948, we need to resolve the above notes.

**Comment by khotilov** [comment]
Fixed the cout's in #3076.
